### ISSUE TITLE
`torch::autograd::grad` broken in libtorch

### ISSUE THREAD

**Issue by dvaCheng** [pytorch/pytorch#113157]
### 🐛 Describe the bug

The example code in pytorch documentation failed to run: [https://pytorch.org/tutorials/advanced/cpp_autograd.html#computing-higher-order-gradients-in-c](https://pytorch.org/tutorials/advanced/cpp_autograd.html#computing-higher-order-gradients-in-c)

Code to repreduce:
```cpp
#include "torch/torch.h"

int main() {
    auto model = torch::nn::Linear(4, 3);

    auto input = torch::randn({3, 4}).requires_grad_(true);
    auto output = model(input);

    // Calculate loss
    auto target = torch::randn({3, 3});
    auto loss = torch::nn::MSELoss()(output, target);

    // Use norm of gradients as penalty
    auto grad_output = torch::ones_like(output);
    auto gradient = torch::autograd::grad({output}, {input}, /*grad_outputs=*/{grad_output}, /*create_graph=*/true)[0];
    auto gradient_penalty = torch::pow((gradient.norm(2, /*dim=*/1) - 1), 2).mean();

    // Add gradient penalty to loss
    auto combined_loss = loss + gradient_penalty;
    combined_loss.backward();

    std::cout << input.grad() << std::endl;
}
```

Error logs:

```
error: C++ exception with description "got 470 tensors and 18446744073709551496 gradients

Exception raised from _make_grads at C:\actions-runner\_work\pytorch\pytorch\builder\windows\pytorch\torch\csrc\autograd\autograd.cpp:55 (most recent call first):

00007FF91177AF4200007FF91177AEE0 c10.dll!c10::Error::Error [<unknown file> @ <unknown line number>]

00007FF91177AACA00007FF91177AA70 c10.dll!c10::detail::torchCheckFail [<unknown file> @ <unknown line number>]

00007FF8BA500BB900007FF8BA500300 torch_cpu.dll!torch::autograd::AnomalyMetadata::store_stack [<unknown file> @ <unknown line number>]

00007FF8BA50101500007FF8BA500FF0 torch_cpu.dll!torch::autograd::grad [<unknown file> @ <unknown line number>]

..."

```

### Versions

```
Libtorch version: 2.1.0+cpu
Is debug build: False
OS: Windows
C++ compiler: cl (msvc 19.34.31935)
Host machine cpu family: x86_64
Host machine cpu: x86_64
```

cc @svekars @carljparker @jbschlosser @peterjc123 @mszhanyi @skyline75489 @nbcsm @vladimir-aubrecht @iremyux @Blackhex @cristianPanaite

**Comment by dvaCheng** [comment]
I believe this bug only occur on Windows. I put the code into a github CI as a unit test, and only Windows build are broken: https://github.com/dvaCheng/DGo/actions/runs/6785576843/job/18444210819

**Comment by dvaCheng** [comment]
Issue solved. It is caused by building with debug option on msvc when using release build of libtorch.
